Utilize `packaging.version.Version`

CasperWA opened the pull request:
Fixes #221 

Since we are currently only supporting Python packages, it makes sense to use the [`packaging.version.Version`](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-scheme) class to parse the latest retrieved version from PyPI. This ensures we can handle the quirks of Python versions.

Addresses the **bold** parts of #217, and the immediate bug there.

Specifically, this addresses the `epoch` and `post-release` segments, adding tests for these as well.

For now, all non `base_version` segments are stripped from the version, meaning only epoch (if present) + release (major.minor.patch) are included from a parsed `Version` instance.

Extra logic is needed to handle and support the remaining `Version` segments, as is also mentioned in #217.

daniel-sintef commented on ci_cd/tasks/update_deps.py:
```diff
@@ -268,9 +281,8 @@ def update_deps(  # pylint: disable=too-many-branches,too-many-locals,too-many-s
             hide=True,
         )
         package_latest_version_line = out.stdout.split(sep="\n", maxsplit=1)[0]
```
The change from re.match to re.search and the pattern update to \S+ (any non-whitespace character) significantly improves the versatility in matching package versions. However, this pattern is now overly broad and may match unexpected strings. A more precise pattern might be warranted to reduce the risk of erroneous matches.

CasperWA replied:
The `\S+` cannot be changed, however, I can change it back to using `match` instead of `search`, which would be fine? Any precision comes later, when validating the two parts.

daniel-sintef commented on ci_cd/tasks/update_deps.py:
```diff
@@ -285,7 +297,7 @@ def update_deps(  # pylint: disable=too-many-branches,too-many-locals,too-many-s
             error = True
             continue
 
-        latest_version: str = match.group("version")
```
Converting the matched version string to a Version object is a good use of the packaging library's capabilities. It allows for more sophisticated handling of version strings. However, there is no error handling if the Version constructor fails (e.g., due to an invalid version string).

daniel-sintef commented on ci_cd/tasks/update_deps.py:
```diff
@@ -365,7 +377,10 @@ def update_deps(  # pylint: disable=too-many-branches,too-many-locals,too-many-s
                     current_version = specifier.version.split(".")
                     break
             else:
```
The handling of the epoch is a bit confusing. The else clause seems inverted; if there is no epoch (epoch == 0), why prepend it to 0.0.0? This seems like a potential error or at least a place where the intent isn't clear. Clearer comments or a refactor for readability might be needed.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -73,7 +74,7 @@ class SemanticVersion(str):
     """
 
-    _REGEX = (
+    _regex = (
```
Example (Line 73-83): The regex pattern _REGEX is used for version parsing. Given the introduction of Version, consider whether relying on custom regex is still the optimal approach. The Version class might render this redundant or less reliable.

CasperWA replied:
This is a regex for semantic versioning, not python versioning. I will rename the variable to denote this.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -171,23 +197,72 @@ def patch(self) -> int:
         """
         return self._build
 
+    @property
```
To simplify the `as_python_version` method for better maintainability, you can refactor it to reduce complexity and improve readability. Here's a proposed approach:

1. **Separate Concerns:** Break down the method into smaller, focused sub-methods. Each sub-method should handle a specific aspect of the version conversion process.

2. **Clear Naming:** Use descriptive names for variables and methods to make the code self-explanatory.

3. **Avoid Deep Nesting:** Reduce nested if-else structures to make the code easier to follow.

Here's a refactored version of your method:

```python
def as_python_version(self, shortened: bool = True) -> Version:
    """Return the Python version as defined by `packaging.version.Version`."""
    if not self.python_version:
        return Version(self.shortened())

    version_parts = [self._get_epoch_segment(), self._get_release_segment(shortened)]
    version_parts.extend(self._get_additional_segments())

    return Version("".join(filter(None, version_parts)))

def _get_epoch_segment(self) -> str:
    """Return the epoch segment of the version."""
    return f"{self.python_version.epoch}!" if self.python_version.epoch != 0 else ""

def _get_release_segment(self, shortened: bool) -> str:
    """Return the release segment of the version."""
    return self.shortened() if shortened else ".".join(
        str(_) for _ in (self.major, self.minor, self.patch)
    )

def _get_additional_segments(self) -> List[str]:
    """Return the pre-release, post-release, development, and local segments."""
    segments = []
    if self._match_release_with_version():
        segments.extend(self._get_pre_post_dev_local_segments())

    return segments

def _match_release_with_version(self) -> bool:
    """Check if the release matches the current version."""
    return (self.major, self.minor, self.patch)[:len(self.python_version.release)] == self.python_version.release

def _get_pre_post_dev_local_segments(self) -> List[str]:
    """Get pre-release, post-release, development, and local segments if they exist."""
    segments = []
    if self.python_version.pre is not None:
        segments.append("".join(str(_) for _ in self.python_version.pre))

    if self.python_version.post is not None:
        segments.append(f".post{self.python_version.post}")

    if self.python_version.dev is not None:
        segments.append(f".dev{self.python_version.dev}")

    if self.python_version.local is not None:
        segments.append(f"+{self.python_version.local}")

    return segments
```

In this refactored version:

- The method `as_python_version` is now streamlined and delegates specific tasks to helper methods.
- Each helper method (`_get_epoch_segment`, `_get_release_segment`, `_get_additional_segments`, `_match_release_with_version`, and `_get_pre_post_dev_local_segments`) handles a distinct part of the version conversion process.
- This approach enhances readability and makes the code easier to maintain and modify in the future.
- The use of list comprehensions and `filter` helps to avoid complex nesting and makes the code more Pythonic.

CasperWA replied:
Sure. And I'd agree if SemanticVersion was all about redoing `packaging.version.Version`, but it is not. Well, it sort of is, but not in this sense. Adding this many utility methods to a class that already has a myriad of methods will most likely result in more confusion.

That said, I do like the last point of using comprehensions and `filter`. I'll try to implement this.

I've already tried to be as verbose as I can in this method, while still keeping it within the method.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -196,7 +271,7 @@ def __str__(self) -> str:
 
     def __repr__(self) -> str:
         """Return the string representation of the object."""
```
The enhanced validation and comparison methods, which now consider Version objects, are a step forward in robustness. However, the increased complexity warrants thorough testing, especially edge cases.

CasperWA replied:
That's why this is also tested.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -292,7 +367,7 @@ def next_version(self, version_part: str) -> SemanticVersion:
 
     def previous_version(
-        self, version_part: str, max_filler: str | int | None = 99
+        self, version_part: str, max_filler: str | int | None = None
```
The logic for determining the previous version seems overly complex and could be simplified. For instance, using a dictionary to map version parts to their decrement logic might make the code more readable and maintainable.

CasperWA replied:
Maybe. But this way, the decrement logic will only ever run for the given version_part. If a dictionary/mapping is created, the whole of its contents will have to be resolved before then choosing which one is needed.

Sometimes readability is valued higher than efficiency and optimized run times, however, the change in readability is minimal here, I'd argue.

It's basically already split into two parts:
1. Value and Type checks for inputs / set defaults.
2. If-branching depending on the three available `version_part` values.

It's the same for `next_version()`, however, since there is no `max_filler` variable to worry about, this part can be left out. Also, the logic for determining a previous version for some version parts are a bit more involved.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -599,28 +666,25 @@ def _ignore_semver_rules(
-        if (
-            len(latest) >= 2
+    if (  # pylint: disable=too-many-boolean-expressions
+        ("major" in semver_rules["version-update"] and latest[0] != current[0])
```
The nested conditional statements are quite complex. Breaking down this logic into smaller, more descriptive helper functions would increase readability and maintainability.

CasperWA replied:
Sure. But I think this is out-of-scope for this PR. What is implemented here is merely a minimization (and simplification) of the code. But I agree that a bit more context could/should be given to this logic.

CasperWA replied:
As an extra comment, I'd say this is (sort of) related to #148, at least the motivation behind creating this issue. I can create another issue to follow up simplification of the logic in this file.

CasperWA replied:
Done. See #222

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -723,16 +787,22 @@ def regenerate_requirement(
     return updated_dependency
 
 
```
The function now accepts SemanticVersion, Version, or str types for latest_version, which is a flexible approach. However, this adds complexity to the function. It's critical to ensure that the function's behavior remains consistent across these different input types.
Suggested Refactoring:
Instead of immediately converting latest_version to SemanticVersion, first check if it's already a Version instance and handle it accordingly. This could reduce the need for multiple conversions and checks.

def update_specifier_set(latest_version, current_specifier_set):
    logger = logging.getLogger(__name__)

    # Refactored handling of version types
    if isinstance(latest_version, Version):
        latest_semantic_version = convert_version_to_semantic_version(latest_version)
    elif isinstance(latest_version, str):
        latest_semantic_version = SemanticVersion(latest_version)
    else:
        latest_semantic_version = latest_version

    new_specifier_set = set(current_specifier_set)
    updated_specifiers = []
    # Rest of the logic remains the same

CasperWA replied:
SemanticVersion supports "wrapping" a Version instance, so no need for this extra logic.

daniel-sintef commented on ci_cd/utils/versions.py:
```diff
@@ -758,8 +828,31 @@ def update_specifier_set(  # pylint: disable=too-many-statements
+            current_version_epochs
+        ):
+            raise UnableToResolve(
+                "The latest version's epoch is smaller than the largest epoch in "
```
should be covered by tests

CasperWA replied:
Is it fine with you to reference #211 and say that's that for now?

daniel-sintef commented:
### Overall general comments: 

- **Areas for Improvement:**
  - **Complexity and Readability:** Some parts, especially the `update_specifier_set` function, exhibit complexity that could hinder maintainability. Breaking down complex logic into smaller functions is advisable.
  - **Error Handling:** The conversion of version strings to `Version` objects lacks error handling, which could lead to issues with invalid version strings.
  - **Overgeneralized Patterns:** The broad regex pattern might lead to unexpected string matches. A more precise pattern could reduce erroneous matches.
  - **Epoch Handling:** The logic surrounding epoch handling is somewhat convoluted and could be made clearer.
  - **Testing Gaps:** Certain new code paths, particularly those handling edge cases, lack test coverage.

#### Specific Suggestions:
- **Refactoring `update_specifier_set`:** Split this function to manage its complexity. For example, handling different version segments (major, minor, patch, epoch) in separate functions could improve clarity.
- **Enhanced Error Handling:** Implement robust error handling for the `Version` constructor to manage invalid version strings.
- **Regex Precision:** Fine-tune the regex pattern used for version matching to avoid capturing unintended strings.
- **Clearer Epoch Logic:** Revise the epoch handling logic for clarity. Perhaps introduce a dedicated method or clear inline documentation to explain the rationale behind the current approach.
- **Increased Test Coverage:** Ensure new logic, particularly around edge cases and error handling, is adequately covered by tests.

daniel-sintef reviewed (commented):
I think this code probably introduces more complexity than it needs to, missing test coverage is also rather worrysome. I think some of the functions need to be refactored a bit mroe

CasperWA commented:
> I think this code probably introduces more complexity than it needs to, missing test coverage is also rather worrysome. I think some of the functions need to be refactored a bit mroe

The missing test coverage I'd consider quite minimal here. However, yes, in general (meaning prior to this PR) the test coverage of parts in `versions.py` is bad, indeed.

Unfortunately, the whole thing is quite complex, and there's a lot of added complexity due to a lot of error handling.
The best way to ensure the complexity is valid is through unit testing, so I'd consider #211 of a higher priority going forward.

CasperWA commented:
Concerning the AI-generated overview, I have addressed all of these points, except the first one: refactoring `update_specifier_set()`. This is indeed a function that should be refactored and tested thoroughly as it's at the core of the whole logic of the `update_deps` task.

CasperWA commented:
> Concerning the AI-generated overview, I have addressed all of these points, except the first one: refactoring `update_specifier_set()`. This is indeed a function that should be refactored and tested thoroughly as it's at the core of the whole logic of the `update_deps` task.

To follow up on this, I have opened #222.
